src/java.base/share/classes/jdk/internal/lang/StableValue.java

```diff
@@ -0,0 +1,442 @@
+        Objects.requireNonNull(list);
+        Objects.checkIndex(index, list.size());
+        Objects.requireNonNull(mapper);
+        return StableAccess.computeIfUnset(list, index, mapper);
+    }
+
+    /**
+     * If no value {@linkplain #isSet() is set} for the StableValue for the provided
+     * {@code key}, attempts to compute and set it as per
+     * {@linkplain StableValue#computeIfUnset(Supplier)} by applying the provided
+     * {@code mapper}.
+     * <p>
+     * This is equivalent to:
+     * {@snippet lang = java:
+     * StableValue<V> stable = map.get(key);
+     * if (stable == null) {
+     *      throw new NoSuchElementException("Unknown key: "+key);
+     * }
+     * if (stable.isSet()) {
+     *     return stable.orThrow();
+     * }
+     * Supplier<V> supplier = () -> mapper.apply(key);
+     * return stable.computeIfUnset(supplier);
+     *}
+     * Except it guarantees that the provided mapper is invoked at most once per distinct
+     * key and might be more resource efficient and performant.
+     *
+     * @param map    from which to get a Stab;eValue
+     * @param key    associated with a StableValue
+     * @param mapper to apply if the StableValue associated with the provided {@code key}
+     *               is {@linkplain StableValue#isSet() not set}
+     * @return the current (pre-existing or computed) value for the provided {@code key}
+     * @param <K> the type of keys maintained by this map
+     * @param <V> the StableValue value type to set
+     * @throws NoSuchElementException if {@linkplain #isError()} is {@code true} for
+     *         the provided {@code key}
+     * @throws NoSuchElementException if the provided {@code map} does not
+     *         {@linkplain Map#containsKey(Object) contain} the provided {@code key}
+     * @throws StackOverflowError if the provided {@code mapper} recursively invoke
+     *         this method for the same {@code key}
+     */
+    static <K, V> V computeIfUnset(Map<K, StableValue<V>> map,
+                                   K key,
+                                   Function<? super K, ? extends V> mapper) {
+        Objects.requireNonNull(map);
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(mapper);
+        return StableAccess.computeIfUnset(map, key, mapper);
+    }
+
+    // Memoized factories
+
+    /**
+     * {@return a new <em>memoized</em> {@linkplain Supplier} backed by an internal
+     * stable value where the provided {@code original} supplier will only be invoked
+     * at most once}
+     * <p>
+     * If the {@code original} suppler invokes the returned supplier recursively,
+     * a StackOverflowError will be thrown when the returned supplier's
+     * {@linkplain Supplier#get()} method is invoked.
+     *
+     * @param original the original Suppler to convert to a memoized Supplier
+     * @param <T>      the memoized type
+     */
+    static <T> Supplier<T> ofSupplier(Supplier<? extends T> original) {
```

Comment:
`ofSupplier` sounds like this method returns a `StableValue` from a `Supplier`. I recommend another name, such as `stableSupplier` or `memoize` to better associate with the method's types.

Reply:
One alternative would be to expose the types `StableList` and `StableMap`. This would allow detection of these types if declared. This would also allow us to expose the methods `computeIfUnset` as instance methods and remove the somewhat strange static counterparts. I agree strict finals are better in the long run.

Reply:
We had other names for the memoized factories before but some people did not like names like `asMemoized`. 

Maybe `ofSupplier` -> `memoizedSupplier` etc. ?